File: frontend/src/services/logger.ts
export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
}

export interface LogEntry {
  timestamp: Date;
  level: LogLevel;
  module: string;
  message: string;
  data?: unknown;
}

const MAX_HISTORY = 1000;

export class Logger {
  readonly module: string;
  private level: LogLevel = LogLevel.DEBUG;
  private history: LogEntry[] = [];

  constructor(module: string) {
    this.module = module;
  }

  setLevel(level: LogLevel): void {
    this.level = level;
  }

  debug(message: string, data?: unknown): void {
    this.log(LogLevel.DEBUG, message, data);
  }

  info(message: string, data?: unknown): void {
    this.log(LogLevel.INFO, message, data);
  }

  warn(message: string, data?: unknown): void {
    this.log(LogLevel.WARN, message, data);
  }

  error(message: string, data?: unknown): void {
    this.log(LogLevel.ERROR, message, data);
  }

  getHistory(): LogEntry[] {
    return [...this.history];
  }

  clearHistory(): void {
    this.history = [];
  }

  private log(level: LogLevel, message: string, data?: unknown): void {
    if (level < this.level) return;

    const entry: LogEntry = {
      timestamp: new Date(),
      level,
      module: this.module,
      message,
      data,
    };

    this.history.push(entry);
    if (this.history.length > MAX_HISTORY) {
      this.history = this.history.slice(-MAX_HISTORY);
    }

    const timestamp = this.formatTimestamp(entry.timestamp);
    const levelStr = `[${LogLevel[level]}]`;
    const moduleStr = `[${this.module}]`;

    const args: unknown[] = [
      `${timestamp} ${levelStr}`,
      moduleStr,
      message,
    ];
    if (data !== undefined) {
      args.push(data);
    }

    switch (level) {
      case LogLevel.DEBUG:
      case LogLevel.INFO:
        console.log(...args);
        break;
      case LogLevel.WARN:
        console.warn(...args);
        break;
      case LogLevel.ERROR:
        console.error(...args);
        break;
    }
  }

  private formatTimestamp(date: Date): string {
    const h = String(date.getHours()).padStart(2, '0');
    const m = String(date.getMinutes()).padStart(2, '0');
    const s = String(date.getSeconds()).padStart(2, '0');
    const ms = String(date.getMilliseconds()).padStart(3, '0');
    return `${h}:${m}:${s}.${ms}`;
  }
}

// Pre-configured loggers for each module
export const createLogger = (module: string): Logger => new Logger(module);

File: backend/StockSim.Engine/Utils/Logger.cs
namespace StockSim.Engine.Utils;

public enum LogLevel
{
    Debug = 0,
    Info = 1,
    Warn = 2,
    Error = 3
}

public record LogEntry(
    DateTime Timestamp,
    LogLevel Level,
    string Module,
    string Message,
    object? Data = null
);

/// <summary>
/// Structured logger for the backend simulation engine.
/// Mirrors the frontend logger design for consistency.
/// </summary>
public class Logger
{
    private const int MaxHistory = 5000;
    public string Module { get; }
    // Default level from STOCKSIM_LOG_LEVEL (e.g. "Warn" for fast headless playtests); Debug otherwise.
    private static readonly LogLevel DefaultLevel =
        Enum.TryParse<LogLevel>(Environment.GetEnvironmentVariable("STOCKSIM_LOG_LEVEL"), ignoreCase: true, out var lvl)
            ? lvl : LogLevel.Debug;
    private LogLevel _minLevel = DefaultLevel;
    private readonly List<LogEntry> _history = new();
    private static readonly object _lock = new();

    public Logger(string module)
    {
        Module = module;
    }

    public void SetLevel(LogLevel level) => _minLevel = level;

    public void Debug(string message, object? data = null) => Log(LogLevel.Debug, message, data);
    public void Info(string message, object? data = null) => Log(LogLevel.Info, message, data);
    public void Warn(string message, object? data = null) => Log(LogLevel.Warn, message, data);
    public void Error(string message, object? data = null) => Log(LogLevel.Error, message, data);

    public IReadOnlyList<LogEntry> GetHistory()
    {
        lock (_lock)
        {
            return _history.ToList().AsReadOnly();
        }
    }

    public void ClearHistory()
    {
        lock (_lock)
        {
            _history.Clear();
        }
    }

    private void Log(LogLevel level, string message, object? data)
    {
        if (level < _minLevel) return;

        var entry = new LogEntry(DateTime.UtcNow, level, Module, message, data);

        lock (_lock)
        {
            _history.Add(entry);
            if (_history.Count > MaxHistory)
            {
                _history.RemoveRange(0, _history.Count - MaxHistory);
            }
        }

        var timestamp = entry.Timestamp.ToString("HH:mm:ss.fff");
        var prefix = $"{timestamp} [{level.ToString().ToUpper()}] [{Module}]";

        var output = data != null
            ? $"{prefix} {message} | {System.Text.Json.JsonSerializer.Serialize(data)}"
            : $"{prefix} {message}";

        switch (level)
        {
            case LogLevel.Debug:
            case LogLevel.Info:
                Console.WriteLine(output);
                break;
            case LogLevel.Warn:
                Console.ForegroundColor = ConsoleColor.Yellow;
                Console.WriteLine(output);
                Console.ResetColor();
                break;
            case LogLevel.Error:
                Console.ForegroundColor = ConsoleColor.Red;
                Console.Error.WriteLine(output);
                Console.ResetColor();
                break;
        }
    }
}
